ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixAggregatedViewClusterInitializer.java

```diff
@@ -0,0 +1,198 @@
+import java.util.concurrent.ExecutionException;
+import org.apache.helix.HelixManager;
+import org.apache.helix.InstanceType;
+import org.apache.helix.PropertyType;
+import org.apache.helix.spectator.RoutingTableProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class represents the startup procedure for registering with helix aggregated view cluster which contains the
+ * cluster information of all colos. Calling the start method will begin the startup procedure in a background thread.
+ * Calling the {@link #join} method let's the main thread wait for startup of this class to either succeed or fail.
+ * This class and its logic is similar to {@link HelixDatacenterInitializer} except that this talks to Helix aggregated
+ * view cluster ZK which has information about entire cluster. So, when {@link ClusterMapConfig#clusterMapUseAggregatedView}
+ * is enabled, {@link HelixClusterManager} would only do cluster initialization once via
+ * {@link HelixAggregatedViewClusterInitializer} instead of calling {@link HelixDatacenterInitializer}s multiple times.
+ */
+class HelixAggregatedViewClusterInitializer {
+  private static final Logger logger = LoggerFactory.getLogger(HelixAggregatedViewClusterInitializer.class);
+  private final CompletableFuture<HelixAggregatedViewClusterInfo> initializationFuture = new CompletableFuture<>();
+  private final ClusterMapConfig clusterMapConfig;
+  private final HelixFactory helixFactory;
+  private final String selfInstanceName;
+  private final DataNodeConfigSourceMetrics dataNodeConfigSourceMetrics;
+  private final HelixClusterManager helixClusterManager;
+  private final Map<String, DcZkInfo> dataCenterToZkAddress;
+
+  /**
+   * @param clusterMapConfig {@link ClusterMapConfig} to help some admin operations
+   * @param dataCenterToZkAddress map of data center to ZK information.
+   * @param helixFactory the {@link HelixFactory} instance to construct managers.
+   * @param selfInstanceName the name of instance on which {@link HelixClusterManager} resides.
+   * @param dataNodeConfigSourceMetrics metrics related to {@link DataNodeConfigSource}.
+   * @param helixClusterManager reference to {@link HelixClusterManager} that creates this class.
+   */
+  HelixAggregatedViewClusterInitializer(ClusterMapConfig clusterMapConfig, Map<String, DcZkInfo> dataCenterToZkAddress,
+      HelixFactory helixFactory, String selfInstanceName, DataNodeConfigSourceMetrics dataNodeConfigSourceMetrics,
+      HelixClusterManager helixClusterManager) {
+    this.clusterMapConfig = clusterMapConfig;
+    this.dataCenterToZkAddress = dataCenterToZkAddress;
+    this.helixFactory = helixFactory;
+    this.selfInstanceName = selfInstanceName;
+    this.dataNodeConfigSourceMetrics = dataNodeConfigSourceMetrics;
+    this.helixClusterManager = helixClusterManager;
+  }
+
+  /**
+   * Begin the startup procedure in a background thread.
+   */
+  public void start() {
+    Utils.newThread(() -> {
```

Comment:
There is only 1 instance of HelixAggregatedViewClusterInitializer, no need to create another thread.  Old approach created multiple initializers and ran them in parallel, then joined the threads.  That's why the old approach used threads.  For this new approach, it creates 1 thread and joins, there is no benefit for using a thread so just remove it.

Reply:
Thank you for pointing it out. I will remove the thread. Had it as a TODO item but forgot about it.